src/compiler/base/main.xsl

```diff
@@ -27,6 +28,32 @@
    <xsl:variable name="initial-document-actual-uri" as="xs:anyURI"
       select="x:document-actual-uri($initial-document)" />
 
+   <!--
+      Accumulators for non-global x:variable
+   -->
+
+   <!-- Push and pop x:variable based on node identity -->
+   <xsl:accumulator name="local:stacked-variables" as="element(x:variable)*" initial-value="()">
+      <xsl:accumulator-rule match="x:scenario/x:variable" select="$value, ." />
+      <xsl:accumulator-rule match="x:scenario" phase="end" select="$value except x:variable" />
```

Comment:
Initially, I misread the `select` attribute, forgetting that `x:variable` is relative to `x:scenario`. Your code is correct, but I wonder if something like `$value except ./child::x:variable` or `<!-- Remove variables declared as children of this scenario -->` would make it a bit harder for that kind of misreading to happen. I haven't worked with accumulators much, which might have contributed to my misunderstanding, so it's fine if you prefer to leave this as is.

Reply:
Added comments and explicit axes to the expressions. (958ead095d433d32b2603cf037871d7aa6dd5d9d)